templates/README.md

```diff
@@ -0,0 +1,83 @@
+# Templates
+
+This directory contains template files that can be expanded and copied over to the configured repositories.
+
+## Header
+
+The [header](header.yml) is prepended to all the files before they are copied to the target repositories.
+
+## Contexts
+
+You can access context information during template expansion(before copy is performed).
+
+### About contexts
+
+Contexts are a way to access information about the repository configuration object and repository GitHub settings. Contexts use the following syntax:
+
+```
+${{{ <context> }}}
+```
+
+| Context name | Type | Description |
+| --- | --- | --- |
+| `config` | `object` | Configuration object because of which the file is being copied. For more information, see [config context](#config-context) |
+| `github` | `object` | Information about the target repository the file is being copied to. For more information, see [github context](#github-context) |
+
+#### `config` context
+
+The `config` context is the configuration object because of which the file is being copied.
+
+| Property name | Type | Always present | Description |
+| --- | --- | --- | --- |
+| `config` | `object` | `true` | The top-level context. | true |
+| `config.files` | `array` | `true` | The files that are being copied. |
+| `config.extra_files` | `array` | `false` | The additional files that are also being copied. |
+| `config.target` | `string` | `true` | The name of the target repository in `{owner}/{repo}` format. |
+| `config.deploy_go` | `boolean` | `false` | Flag controling if Go specific setup should be applied to the repository. |
+| `config.deploy_versioning` | `boolean` | `false` | Flag controling if `versions.json` file should be deployed to the repository. |
+| `config.dist` | `string` | `false` | The name of the distribution built from the repository. |
+
+#### `github` context
+
+The `github` context contains information about the target repository the file is being copied to.
+
+| Property name | Type | Always present | Description |
+| --- | --- | --- | --- |
+| `github` | `object` | `true` | The top-level context. |
+| `github.default_branch` | `string` | `true` | The name of the default branch of the target repository. |
+
+### Examples
+
+#### Context container
+
+```json
+{
+    "config": {
+        "example": {
+            "greeting": "Hello"
+        },
+        "files": [],
+        "extra_files": [".github/workflows/example.yml"],
+        "target": "protocol/.github-test-target"
+    },
+    "github": {
+        "default_branch": "master"
```

Comment:
Why is this a config value? This can be easily obtained from the github context.

Reply:
I don't think [github context](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context) has the name of the default branch. However, even if it did, we might want to use that information prior to the context being available, just as we did here for example: https://github.com/protocol/.github/blob/dfff3e56fa8ec5ecd0ce0188489e62f94a1be9fe/templates/.github/workflows/releaser.yml#L5

And it looks like the JS workflow will need it soon too: https://github.com/protocol/.github/blob/5cec3f5077e1d6ce52c17c73c85460793085feaf/templates/.github/workflows/js-test-and-release.yml#L5

Reply:
Right, it's not available in `if` conditions, so it makes sense to _somehow_ replace this.
I'm wondering though if configuring it in a config file is the right way to go. The copy workflow now determines the name of the default branch on the fly, which makes for a smoother workflow when repos rename their default branch.

Reply:
Oh yes, I do completely agree and that's why I didn't change it. The `github` template context object is still created on the fly during `copy-workflow`'s execution. See https://github.com/protocol/.github/blob/406b18365f417643ce1aaf92916f3b2a49868f37/.github/workflows/copy-workflow.yml#L110-L113

It's only the `config` template context object that comes from the "parsed JSON"/whatever the input to the `copy-workflow` was.

Reply:
What does setting the `github.default_branch` do then? I'm confused now :)

Reply:
This discussion is happening under a line from an example of a [context container](https://github.com/protocol/.github/blob/486f183d83a80dd1a48589b8bfbf1672f8167064/templates/README.md#context-container) - an object that we can refer to during template expansion.

If I write a template that has `${{{ some.json.path }}}` somewhere in its content, then that string will be replaced with the output of `echo "$CONTEXT_CONTAINER" | jq -jc '.some.json.path'` - the value of `some.json.path` inside the context container.

The context container, at least for now, has 2 keys - `config` and `github`.

The `github`'s value is an object that is created on the fly during the execution of the `copy-workflow`. Currently, it has one key - `default_branch` - but we can add more in the future if we want to.

The `config`'s value is an object that was passed to `copy-workflow` as [the input](https://github.com/protocol/.github/blob/406b18365f417643ce1aaf92916f3b2a49868f37/.github/workflows/copy-workflow.yml#L21). So it's, for example, a merge of defaults https://github.com/protocol/.github/blob/406b18365f417643ce1aaf92916f3b2a49868f37/configs/go.json#L2-L13 and a specific repository config https://github.com/protocol/.github/blob/406b18365f417643ce1aaf92916f3b2a49868f37/configs/go.json#L15

Reply:
Got it, so the context container is generated automatically. Maybe that's worth spelling out explicitly in the README?

Reply:
Good idea, let me add this :)

Reply:
Done! I also added and example of JSON config and expanded template to the examples - https://github.com/protocol/.github/pull/262/commits/676e4905ddea2065df45b02b515a75d7179af45a.